// ka-table/Utils/DataUtils.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import { getField, getLastField, getLastFieldParents } from './ColumnUtils';
export var getParentValue = function (rowData, fieldParents, sameObj) {
    if (sameObj === void 0) { sameObj = false; }
    var parentValue = fieldParents.reduce(function (previousValue, currentValue) {
        var result = (previousValue && previousValue[currentValue]);
        return result !== undefined ? result : undefined;
    }, rowData);
    return parentValue ? __assign({}, parentValue) : undefined;
};
export var createObjByFields = function (fieldParents, field, value) {
    var parentValue = {};
    if (fieldParents.length) {
        fieldParents.reduce(function (previousValue, currentItem, currentIndex) {
            var lastObj = {};
            previousValue[currentItem] = lastObj;
            if (currentIndex === (fieldParents.length - 1)) {
                lastObj[field] = value;
            }
            return lastObj;
        }, parentValue);
    }
    else {
        parentValue[field] = value;
    }
    return __assign({}, parentValue);
};
export var getValueByColumn = function (rowData, column) {
    return getValueByField(rowData, getField(column));
};
export var getValueByField = function (rowData, field) {
    var o = __assign({}, rowData);
    var names = field.split('.');
    for (var i = 0, n = names.length; i < n; ++i) {
        var k = names[i];
        if (k in o) {
            o = o[k];
        }
        else {
            return;
        }
    }
    return o;
};
var _replaceValueForField = function (rowData, field, newValue, fieldParents) {
    var result = __assign({}, rowData);
    if (fieldParents && fieldParents.length) {
        var parentValue = getParentValue(result, fieldParents) || {};
        parentValue[field] = newValue;
        var parentsOfParent = __spreadArrays(fieldParents);
        var parentFieldName = parentsOfParent.pop();
        result = _replaceValueForField(result, parentFieldName, parentValue, parentsOfParent);
    }
    else {
        result[field] = newValue;
    }
    return result;
};
export var replaceValue = function (rowData, column, newValue) {
    return replaceValueForField(rowData, getField(column), newValue);
};
export var replaceValueForField = function (rowData, field, newValue) {
    return _replaceValueForField(rowData, getLastField(field), newValue, getLastFieldParents(field));
};
