// ka-table/Utils/PropsUtils.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import { isFunction } from 'util';
import { filterData, searchData } from './FilterUtils';
import { getGroupedData } from './GroupUtils';
import { getPageData, getPagesCount } from './PagingUtils';
import { sortData } from './SortUtils';
import { convertToColumnTypes } from './TypeUtils';
export var extendProps = function (childElementAttributes, childProps, childComponent) {
    var resultProps = childElementAttributes;
    var childCustomAttributes = childComponent && childComponent.elementAttributes && childComponent.elementAttributes(childProps);
    if (childCustomAttributes) {
        var dispatch = childProps.dispatch;
        resultProps = mergeProps(childElementAttributes, childProps, childCustomAttributes, dispatch);
    }
    return resultProps;
};
var emptyFunc = function () { };
export var mergeProps = function (childElementAttributes, childProps, childCustomAttributes, dispatch) {
    var customPropsWithEvents = {};
    var _loop_1 = function (prop) {
        if (childCustomAttributes.hasOwnProperty(prop)) {
            var propName = prop;
            var propValue_1 = childCustomAttributes[propName];
            var baseFunc_1 = childElementAttributes[propName] || emptyFunc;
            if (isFunction(propValue_1)) {
                customPropsWithEvents[prop] = function (e) {
                    propValue_1(e, {
                        baseFunc: baseFunc_1,
                        childElementAttributes: childElementAttributes,
                        childProps: childProps,
                        dispatch: dispatch,
                    });
                };
            }
        }
    };
    for (var prop in childCustomAttributes) {
        _loop_1(prop);
    }
    var mergedResult = __assign(__assign(__assign(__assign({}, childElementAttributes), childCustomAttributes), customPropsWithEvents), { className: (childElementAttributes.className || '') + " " + (childCustomAttributes.className || ''), style: __assign(__assign({}, childCustomAttributes.style), childElementAttributes.style) });
    return mergedResult;
};
var _filterData = function (props) {
    var extendedFilter = props.extendedFilter, columns = props.columns, searchText = props.searchText, search = props.search;
    var _a = props.data, data = _a === void 0 ? [] : _a;
    data = __spreadArrays(data);
    data = extendedFilter ? extendedFilter(data) : data;
    data = searchText ? searchData(columns, data, searchText, search) : data;
    data = convertToColumnTypes(data, columns);
    data = filterData(data, columns);
    return data;
};
export var getData = function (props) {
    var columns = props.columns, groups = props.groups, groupsExpanded = props.groupsExpanded, paging = props.paging;
    var _a = props.data, data = _a === void 0 ? [] : _a;
    data = __spreadArrays(data);
    data = _filterData(props);
    data = sortData(columns, data);
    var groupedColumns = groups ? columns.filter(function (c) { return groups.some(function (g) { return g.columnKey === c.key; }); }) : [];
    var groupedData = groups ? getGroupedData(data, groups, groupedColumns, groupsExpanded) : data;
    data = getPageData(groupedData, paging);
    return data;
};
export var prepareTableOptions = function (props) {
    var groups = props.groups, paging = props.paging;
    var columns = props.columns;
    var groupedData = getData(props);
    var groupColumnsCount = 0;
    var groupedColumns = [];
    if (groups) {
        groupColumnsCount = groups.length;
        groupedColumns = columns.filter(function (c) { return groups.some(function (g) { return g.columnKey === c.key; }); });
        columns = columns.filter(function (c) { return !groups.some(function (g) { return g.columnKey === c.key; }); });
    }
    var pagesCount = 1;
    if (paging && paging.enabled) {
        pagesCount = getPagesCount(_filterData(props), paging);
    }
    return {
        columns: columns,
        groupColumnsCount: groupColumnsCount,
        groupedColumns: groupedColumns,
        groupedData: groupedData,
        pagesCount: pagesCount
    };
};
export var getDraggableProps = function (key, dispatch, actionCreator, draggedClass, dragOverClass) {
    var count = 0;
    return {
        draggable: true,
        onDragStart: function (event) {
            count = 0;
            event.dataTransfer.setData('ka-draggableKeyValue', JSON.stringify(key));
            event.currentTarget.classList.add(draggedClass);
            event.dataTransfer.effectAllowed = 'move';
        },
        onDragEnd: function (event) {
            event.currentTarget.classList.remove(draggedClass);
        },
        onDrop: function (event) {
            event.currentTarget.classList.remove(dragOverClass);
            var draggableKeyValue = JSON.parse(event.dataTransfer.getData('ka-draggableKeyValue'));
            dispatch(actionCreator(draggableKeyValue, key));
        },
        onDragEnter: function (event) {
            count++;
            if (!event.currentTarget.classList.contains(dragOverClass)) {
                event.currentTarget.classList.add(dragOverClass);
            }
            event.preventDefault();
        },
        onDragLeave: function (event) {
            count--;
            if (count === 0) {
                event.currentTarget.classList.remove(dragOverClass);
            }
        },
        onDragOver: function (event) {
            if (!event.currentTarget.classList.contains(dragOverClass)) {
                event.currentTarget.classList.add(dragOverClass);
            }
            event.preventDefault();
        }
    };
};
