// ka-table/actionCreators.js
import { ActionType } from './enums';
export var updateFilterRowValue = function (columnKey, filterRowValue) { return ({
    columnKey: columnKey,
    filterRowValue: filterRowValue,
    type: ActionType.UpdateFilterRowValue,
}); };
export var updateFilterRowOperator = function (columnKey, filterRowOperator) { return ({
    columnKey: columnKey,
    filterRowOperator: filterRowOperator,
    type: ActionType.UpdateFilterRowOperator,
}); };
export var updateEditorValue = function (rowKeyValue, columnKey, value) { return ({
    columnKey: columnKey,
    rowKeyValue: rowKeyValue,
    type: ActionType.UpdateEditorValue,
    value: value,
}); };
export var updateCellValue = function (rowKeyValue, columnKey, value) { return ({
    columnKey: columnKey,
    rowKeyValue: rowKeyValue,
    type: ActionType.UpdateCellValue,
    value: value,
}); };
export var updateSortDirection = function (columnKey) { return ({
    columnKey: columnKey,
    type: ActionType.UpdateSortDirection,
}); };
export var closeEditor = function (rowKeyValue, columnKey) { return ({
    columnKey: columnKey,
    rowKeyValue: rowKeyValue,
    type: ActionType.CloseEditor,
}); };
export var deleteRow = function (rowKeyValue) { return ({
    rowKeyValue: rowKeyValue,
    type: ActionType.DeleteRow,
}); };
export var deselectAllRows = function () { return ({
    type: ActionType.DeselectAllRows,
}); };
export var deselectRow = function (rowKeyValue) { return ({
    rowKeyValue: rowKeyValue,
    type: ActionType.DeselectRow,
}); };
export var openEditor = function (rowKeyValue, columnKey) { return ({
    columnKey: columnKey,
    rowKeyValue: rowKeyValue,
    type: ActionType.OpenEditor,
}); };
export var search = function (searchText) { return ({
    searchText: searchText,
    type: ActionType.Search,
}); };
export var selectAllRows = function () { return ({
    type: ActionType.SelectAllRows,
}); };
export var selectSingleRow = function (rowKeyValue) { return ({
    rowKeyValue: rowKeyValue,
    type: ActionType.SelectSingleRow,
}); };
export var selectRow = function (rowKeyValue) { return ({
    rowKeyValue: rowKeyValue,
    type: ActionType.SelectRow,
}); };
export var selectRowsRange = function (rowKeyValueFrom, rowKeyValueTo) { return ({
    rowKeyValueFrom: rowKeyValueFrom,
    rowKeyValueTo: rowKeyValueTo,
    type: ActionType.SelectRowsRange,
}); };
export var updateGroupsExpanded = function (groupKey) { return ({
    groupKey: groupKey,
    type: ActionType.UpdateGroupsExpanded,
}); };
export var updateData = function (data) { return ({
    data: data,
    type: ActionType.UpdateData,
}); };
export var showLoading = function (text) { return ({
    text: text,
    type: ActionType.ShowLoading,
}); };
export var hideLoading = function () { return ({
    type: ActionType.HideLoading,
}); };
export var showNewRow = function () { return ({
    type: ActionType.ShowNewRow,
}); };
export var hideNewRow = function () { return ({
    type: ActionType.HideNewRow,
}); };
export var showDetailsRow = function (rowKeyValue) { return ({
    rowKeyValue: rowKeyValue,
    type: ActionType.ShowDetailsRow,
}); };
export var hideDetailsRow = function (rowKeyValue) { return ({
    rowKeyValue: rowKeyValue,
    type: ActionType.HideDetailsRow,
}); };
export var saveNewRow = function (rowKeyValue, settings) { return ({
    rowKeyValue: rowKeyValue,
    validate: settings && settings.validate,
    type: ActionType.SaveNewRow,
}); };
export var openRowEditors = function (rowKeyValue) { return ({
    rowKeyValue: rowKeyValue,
    type: ActionType.OpenRowEditors,
}); };
export var closeRowEditors = function (rowKeyValue) { return ({
    rowKeyValue: rowKeyValue,
    type: ActionType.CloseRowEditors,
}); };
export var saveRowEditors = function (rowKeyValue, settings) { return ({
    rowKeyValue: rowKeyValue,
    validate: settings && settings.validate,
    type: ActionType.SaveRowEditors,
}); };
export var updateRow = function (rowData) {
    return {
        type: ActionType.UpdateRow,
        rowData: rowData,
    };
};
export var updatePageIndex = function (pageIndex) { return ({
    pageIndex: pageIndex,
    type: ActionType.UpdatePageIndex,
}); };
export var updatePagesCount = function (pagesCount) { return ({
    pagesCount: pagesCount,
    type: ActionType.UpdatePagesCount,
}); };
export var resizeColumn = function (columnKey, width) { return ({
    type: ActionType.ResizeColumn,
    columnKey: columnKey,
    width: width,
}); };
export var reorderRows = function (rowKeyValue, targetRowKeyValue) { return ({
    type: ActionType.ReorderRows,
    rowKeyValue: rowKeyValue,
    targetRowKeyValue: targetRowKeyValue,
}); };
