// ka-table/enums.js
export var DataType;
(function (DataType) {
    DataType["Boolean"] = "boolean";
    DataType["Date"] = "date";
    DataType["Number"] = "number";
    DataType["Object"] = "object";
    DataType["String"] = "string";
})(DataType || (DataType = {}));
export var EditingMode;
(function (EditingMode) {
    EditingMode["None"] = "none";
    EditingMode["Cell"] = "cell";
})(EditingMode || (EditingMode = {}));
export var ActionType;
(function (ActionType) {
    ActionType["CloseEditor"] = "CloseEditor";
    ActionType["CloseRowEditors"] = "CloseRowEditors";
    ActionType["DeleteRow"] = "DeleteRow";
    ActionType["DeselectAllRows"] = "DeselectAllRows";
    ActionType["DeselectRow"] = "DeselectRow";
    ActionType["HideDetailsRow"] = "HideDetailsRow";
    ActionType["HideLoading"] = "HideLoading";
    ActionType["HideNewRow"] = "HideNewRow";
    ActionType["OpenEditor"] = "OpenEditor";
    ActionType["OpenRowEditors"] = "OpenRowEditors";
    ActionType["ReorderRows"] = "ReorderRows";
    ActionType["ResizeColumn"] = "ResizeColumn";
    ActionType["SaveNewRow"] = "SaveNewRow";
    ActionType["SaveRowEditors"] = "SaveRowEditors";
    ActionType["ScrollTable"] = "ScrollTable";
    ActionType["Search"] = "Search";
    ActionType["SelectAllRows"] = "SelectAllRows";
    ActionType["SelectRow"] = "SelectRow";
    ActionType["SelectRowsRange"] = "SelectRowsRange";
    ActionType["SelectSingleRow"] = "SelectSingleRow";
    ActionType["ShowDetailsRow"] = "ShowDetailsRow";
    ActionType["ShowLoading"] = "ShowLoading";
    ActionType["ShowNewRow"] = "ShowNewRow";
    ActionType["UpdateCellValue"] = "UpdateCellValue";
    ActionType["UpdateData"] = "UpdateData";
    ActionType["UpdateEditorValue"] = "UpdateEditorValue";
    ActionType["UpdateFilterRowOperator"] = "UpdateFilterRowOperator";
    ActionType["UpdateFilterRowValue"] = "UpdateFilterRowValue";
    ActionType["UpdateGroupsExpanded"] = "UpdateGroupsExpanded";
    ActionType["UpdatePageIndex"] = "UpdatePageIndex";
    ActionType["UpdatePagesCount"] = "UpdatePagesCount";
    ActionType["UpdateRow"] = "UpdateRow";
    ActionType["UpdateSortDirection"] = "UpdateSortDirection";
    ActionType["UpdateVirtualScrolling"] = "UpdateVirtualScrolling";
})(ActionType || (ActionType = {}));
export var KeyboardEnum;
(function (KeyboardEnum) {
    KeyboardEnum[KeyboardEnum["Esc"] = 27] = "Esc";
    KeyboardEnum[KeyboardEnum["Enter"] = 13] = "Enter";
})(KeyboardEnum || (KeyboardEnum = {}));
export var SortDirection;
(function (SortDirection) {
    SortDirection["Ascend"] = "ascend";
    SortDirection["Descend"] = "descend";
})(SortDirection || (SortDirection = {}));
export var SortingMode;
(function (SortingMode) {
    SortingMode["None"] = "none";
    SortingMode["Single"] = "single";
})(SortingMode || (SortingMode = {}));
export var FilteringMode;
(function (FilteringMode) {
    FilteringMode["None"] = "none";
    FilteringMode["FilterRow"] = "filterRow";
})(FilteringMode || (FilteringMode = {}));
export var FilterOperatorName;
(function (FilterOperatorName) {
    FilterOperatorName["Equal"] = "=";
    FilterOperatorName["MoreThan"] = ">";
    FilterOperatorName["LessThan"] = "<";
    FilterOperatorName["MoreThanOrEqual"] = ">=";
    FilterOperatorName["LessThanOrEqual"] = "<=";
    FilterOperatorName["Contains"] = "contains";
    FilterOperatorName["IsEmpty"] = "IsEmpty";
    FilterOperatorName["IsNotEmpty"] = "IsNotEmpty";
})(FilterOperatorName || (FilterOperatorName = {}));
